src/main/java/org/checkerframework/specimin/Slicer.java

```diff
@@ -0,0 +1,137 @@
+package org.checkerframework.specimin;
+
+import com.github.javaparser.StaticJavaParser;
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.ast.ImportDeclaration;
+import com.github.javaparser.ast.Node;
+import com.github.javaparser.ast.PackageDeclaration;
+import com.github.javaparser.resolution.Resolvable;
+import com.github.javaparser.resolution.UnsolvedSymbolException;
+import com.github.javaparser.resolution.declarations.ResolvedReferenceTypeDeclaration;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayDeque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import org.checkerframework.specimin.unsolved.UnsolvedSymbolAlternates;
+import org.checkerframework.specimin.unsolved.UnsolvedSymbolGenerator;
+
+public class Slicer {
+  /** The slice of nodes. */
+  private final Set<Node> slice = new HashSet<>();
+
+  /**
+   * The slice of unsolved symbol definitions. These values need not be unique; the map is provided
+   * for simple lookups when adding new symbols.
```

Comment:
I'm concerned that these values don't need to be unique - maps don't really make sense unless the keys aren't going to overlap. Moreover, I'd generally expect that unsolved symbol names are unique (at least, within a single compilation unit): if not, the original Java code wouldn't compile, either - at least, assuming that the keys here are fully-qualified names. Are they?

Reply:
> However, using sets as the keys would prompt a new SomeClass, with only FQN org.example.SomeClass, to be generated

I don't think this is necessarily a problem - we could replace the key with an updated one when we discover this constraint, no?

> This is an edge edge case, so we may never encounter this. Do you think it's worth accounting for?

I think it's a good idea to try to get this right the first time, since we're rewriting the core of the analysis. So, let's try to make sure we've thought of and accounted for all of the edge cases.

Reply:
> My understanding is that you're suggesting a map of type Map<Set<String>, UnsolvedSymbol>, right?

I was imagining a custom data structure for the key that's backed by a string set but has its own looser equality method. Regardless, I think we're on the same page with respect to what this needs to handle; the exact way that you implement it is less important. You should do what you find most intuitive.